dali/operators/input/video_input.cc

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/input/video_input.h"
+#include <memory>
+#include <vector>
+
+namespace dali {
+
+/// By definition, the input batch size of this Operator is always 1.
+constexpr int input_batch_size = 1;
+
+
+template<>
+bool VideoInput<CPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                       const Workspace &ws) {
+  if (!valid_) {
+    InputOperator<CPUBackend>::HandleDataAvailability();
+    TensorList<CPUBackend> encoded_videos;
+    encoded_videos.set_pinned(device_id_ != CPU_ONLY_DEVICE_ID);
+    frames_decoders_.resize(input_batch_size);
+    auto &thread_pool = ws.GetThreadPool();
+    this->ForwardCurrentData(encoded_videos, thread_pool);
+
+    // Creating FramesDecoders
+    auto sample = encoded_videos[0];
+    auto data = reinterpret_cast<const char *>(sample.data<uint8_t>());
+    size_t size = sample.shape().num_elements();
+    frames_decoders_[0] = std::make_unique<FramesDecoder>(data, size, false);
+
+    assert(output_descs_.empty());
+    DetermineOutputDescs(static_cast<int>(frames_decoders_[0]->NumFrames()));
+
+    // This has to be done for every video file, since we need to know the shape of the frames.
+    if (last_sequence_policy_ == "pad") {
+      InitializePadValue(0);
+    }
+
+    valid_ = true;
+  }
+  output_desc.resize(1);
+  output_desc[0] = output_descs_.front();
+  output_descs_.pop_front();
+  return true;
+}
+
+
+template<>
+void VideoInput<CPUBackend>::RunImpl(Workspace &ws) {
+  auto &output = ws.Output<CPUBackend>(0);
+  bool full_sequence;
+  for (int64_t s = 0; s < output.num_samples(); s++) {
+    auto pad_value =
+            last_sequence_policy_ == "pad" ? std::optional<SampleView<CPUBackend>>(GetPadFrame())
+                                           : std::nullopt;
+    full_sequence = DecodeFrames(output[s], 0, frames_per_sequence_, pad_value);
+    if (!full_sequence) {
+      break;
+    }
+  }
+  if (!full_sequence || frames_decoders_[0]->NextFrameIdx() == -1) {
+    Invalidate();
+  }
+}
+
+
+DALI_SCHEMA(experimental__inputs__Video)
+                .DocStr(
+                        R"code(
+Streams and decodes a video from a memory buffer. To be used with long and high resolution videos.
+
+Returns a batch of sequences of frames, with the layout: ``(N, F, H, W, C)``, where:
```

Comment:
Do you actually set the layout to the output tensor?

Reply:
I'd prefer to leave the description of the batch in the layout too, unless @jantonguirao thinks it shouldn't be here?

Added the test for the layout in `test_video.py`.

Reply:
What worries me is that the TL layout (that is printed in python with `__repr__` will say FHWC), which may be confusing. If you were to pass the output to another dali python you would specify the layout as "FHWC" too, right? Instead of N I would just mention the number of samples/sequences in the batch. Having said that, I am not going to insist, just leaving my concerncs here.

Reply:
Actually, leaving the batch information in the layout is a bit misleading, since we omit it purposedly in the rest of DALI documentation. We've seen situations where the user tried to use the batch information in the layout and had to explain that in DALI the batch is implicit and we do not handle batch information explicitly in the pipeline.. So yes, I agree with Kamil that the batch information should be omitted here.

Reply:
BTW, I can see that readers.video has the batch dimension in the description. I prefer if we stick to omitting the layout, otherwise the users will start fetching/setting/etc the channel dimension and will be confused. However, you may choose to leave it as is to be consistent with video reader and I won't be insisting more.

Reply:
I went over everything again, thought it through again and @stiepan 's point makes it. When specifying the layout in another operator, user would put `FHWC`, not `NFHWC`. I'll omit the `N` in the documentation.